src/ai/ai.cpp

```diff
@@ -265,50 +458,241 @@ static bool ai_has_mutual_enemy(sys::state& state, dcon::nation_id from, dcon::n
 	return false;
 }
 
-constexpr inline float ally_overestimate = 2.f;
-
 bool ai_will_accept_alliance(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
-	if(!state.world.nation_get_ai_is_threatened(target))
+	auto rel = nations::get_diplomatic_relation(state, target, from);
+
+	if(military::are_at_war(state, target, from))
 		return false;
 
 	// Has not surpassed infamy limit
 	if(state.world.nation_get_infamy(target) >= state.defines.badboy_limit * 0.75f)
 		return false;
 
+	// vassals always ally liege
```

Comment:
vassals don't ally their liege in PA because they are automatically called into wars with their overlord regardless of alliance status. It is thus easier to leave them as not allied